### ISSUE TITLE
Add Shapefiles as a provider

### ISSUE THREAD

**Issue by fawick** [terranodo/tegola#160]
I wonder whether it would be of interest for tegola to have a `mvt.Provider` for ESRI shapefiles.

@jonas-p created a [Go package for reading (and writing) ESRI Shapefiles])(https://github.com/jonas-p/go-shp) for which I am a secondary maintainer and which we are currently getting into a more mature shape. It's under MIT license, same as tegola. We already support reading Shapefiles and their indices and property databases from ZIP archives.

**Comment by ARolek** [comment]
@fawick absolutely. We have been focused on plumbing PostGIS end to end, but as you have discovered we stubbed out a `mvt.Provider` interface so we could support multiple backends.

I did a quick look at the shapefile package and I didn't see any documentation about reading the indexes. I'm assuming these are spatial indexes. Can you point me to the right spot?

**Comment by fawick** [comment]
Sorry, my sentence about the indices was misleading. What I meant to say is that the package  assumes a ZIP will bundle `.shp` (geometry), `.shx` (index) and `.dbf` (properties) are bundled in the ZIP, it does not yet read the `.shx` directly, yet.

The `.shx` index file is not a spatial index. It's simply a list of offset addresses into the geometry file (the `.shp`) that allows random access for features without reading the file from the start.

**Comment by ARolek** [comment]
@fawick got it.To query the shapefile efficiently we're going to need some sort of spatial index. We can either read all the features at startup time and build the index in memory or create an index to work with. I did a bit of research and it looks like the .sbn file for shapefiles is where the spatial index would live. According to the [wikipedia article](https://en.wikipedia.org/wiki/Shapefile#Shapefile_spatial_index_format_.28.sbn.29) this is not documented but has been reverse engineered by the open source community. I found a PDF of the file layout - [sbn_format.pdf](https://github.com/terranodo/tegola/files/1423058/sbn_format.pdf)
. I think we would need some sort of spatial index to work with if we're going to fetch features in a performant manner.

Is that something you would consider adding to the go-shp package?

**Comment by fawick** [comment]
Thanks for the PDF, @ARolek! I agree that this is something we should consider for go-shp. 
At the moment I have only a very limited understanding on the algorithms needed to compute a performant spatial index so it can take bit of time until I am up to speed. Using https://github.com/dhconnelly/rtreego would be one of my first ideas.

As the index computation might be interesting for WKB / GeoJSON, too, should become a package in its own rights which operates on the abstract Geometry types?

**Comment by ARolek** [comment]
@fawick you nailed it regarding the index computations. GeoJSON is going to absolutely need this in order to be a tegola provider. WKB would benefit from this as well, though the current use cases (GeoPackage and PostGIS) both have indexes implemented at the provider layer.

>As the index computation might be interesting for WKB / GeoJSON, too, should it become a package in its own rights which operates on the abstract Geometry types?

This opens up a great discussion. We have been working on setting up the [go-spatial](https://github.com/go-spatial) org. Our first step has been establishing core geometry types and interfaces to help drive interoperability between various packages. A package / sub-package for generating various spatial indexes would be a great addition and something that the various encoders / decoders could leverage.

**Comment by murphy214** [comment]
Now when we say index do we mean like a plain rtree or a more domain specific structure being like a map[m.TileID][]Features, this has easy conversion to bounds and back and is more expressive for what the use of the index would probably be. For example lets say you have a set of features that are cleanly within (clipped) tile (0,0,0) you could then get the children of that tile and recursively clip down while knowing their only 4 tiles in the next zoom for each tileid []feature set.

You can then start to make little assumptions by using the children tileid bbs against the a features bb, doing logical assumptions like if the features bounding box is the same size as the previous tileid (parent) and we only have 4 coordinates in the first polygon ring we can then send through 4 rectangle polygons for each child as it is a square. Another assumption is if a features bb exists entirely in a child bb then send the entire feature through to the single child. These little assumptions reduce load on your clipper, quite a bit at lower zooms.

I'm not sure if this is what you had in mind for index, but this structure will have to implemented somewhere in the pipeline I'd think. What type of index did you guys have in mind?

**Comment by ARolek** [comment]
@murphy214 we're rounding out the clipping and make valid part of tegola right now. The release candidate for v0.4.0 should be cut in the next day or so. 

The discussion here is more about performant bbox queries. Tegola is going to send a `tegola.Tile` struct in when it requests an `MVTLayer` from a provider. That tile struct will have z, x, y values and has a method for generating a BoundingBox for the tile using the z, x, y values. The underlying data provider should use this bounding box to lookup which features intersect with it. I'm hoping the `go-shp` package can provide quick lookups using a spatial index which will return unclipped features. Tegola will then handle the rest of the processing.

**Comment by murphy214** [comment]
Check out this gist I made, is this sort of what your after? I used a library of mine for tile to zoom stuff and a geojson library that I build a lot of my stuff on top of. https://gist.github.com/murphy214/6876dd66ccb1d498a2769815564d35fb

**Comment by ARolek** [comment]
@murphy214 thanks for the example code! Reading through it, this looks like a way to check for intersecting bounding boxes but requires iterating through each feature. You are doing it in parallel which will be faster, but it still requires iterating everything. Am I missing something? Have you built out an r-tree index? I think an r-tree would essentially include a step before this step which would minimize the comparisons necessary.

**Comment by murphy214** [comment]
While it does require you to iterate through every feature this process only needs to be done once NOT every request and would replace the dataset entirely furthermore could be done during ingestion like reading in a shp file or geojson. It also uses pointers so the only memory bloat your getting is minimal from so many map entries. (which usually isn't that bad)

So you have this index sitting as your provider, a request comes in for a tile X,Y,Z you already have a map[TileID{X,Y,Z}][]feature structure. In the example code try iterating through each k,v of the tilemap to see what I mean.

**So when a request comes in for a tileid (X,Y,Z) we do something like:**
* features that need clipped for this tileid = map_index[tileid]
* clip(features)
* make tile

To be honest, I'm not sure how an R-Tree would eliminate comparisons to be honest, this is really just a super concise implementation of an R-Tree for the problem domain were addressing but I'm definitely open to hearing suggestions!

**Comment by ARolek** [comment]
@murphy214 I did some additional research on this and apparently an r-tree index allows the query to traverse the set of bounding boxes in a similar fashion to a binary tree. The tree lookup will be substantially faster than iterating every feature. I have not played with any of the implementations but here are two examples in go: [rtreego](https://github.com/dhconnelly/rtreego), [rtree](https://github.com/tidwall/rtree)

**Comment by murphy214** [comment]
"The tree lookup will be substantially faster than iterating every feature."

A query doesn't iterate through every feature, the process that iterates through every feature simply builds the index. All a query has to do in my index is get the tile x,y,z then look it up in a map. If I recall a hash table is substantially faster than a binary tree, especially considering your looking a hash table once for each query rather than several times for each r-tree.

**Comment by ARolek** [comment]
@murphy214 I understand your approach builds out the bounding box indexes at initialization. I missed the fact that you're using the tile{x,y,x} as the map key. That would be a quick lookup but if I understand this correctly you're index would need to pre calculating every bbox intersection for every tile / feature combination on init. Slippy map tiles have the possibility of producing [billions of tiles](http://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Zoom_levels) at the higher zoom levels. This approach has the possibility of generating that many index entries as well. 

I have not implemented an r-tree index, but my understanding is that we would have all the bounding boxes of the features indexed and then we could query the r-tree index with a bounding box to find feature intersections. This would require keeping track of all bounding boxes, where as your approach keeps track of tile x,y,z : feature pairs. Interesting tradeoffs to consider.

**Comment by murphy214** [comment]
You could consider stopping at like zoom 14 or 16 for an index, as thats really quite a high resolution for clipping anyway keep in mind youd still be clipping around the high tile zoom square you'd just be looking up at zoom 16 for anything above 16. Although this would introduce confusion in the api that would have to dealt with (i.e. having a different x,y,z other than the tile your making) but other than that its a good way to drastically reduce the index size. I checked the relative sizes of each in linear distance diagonally size 14 is like 2 miles diagonally while size 16 is like half a mile.
